drivers/network/netio/netio.c

```diff
@@ -0,0 +1,860 @@
+    PSOCKADDR LocalAddress,
+    PSOCKADDR RemoteAddress,
+    ULONG Flags,
+    PVOID SocketContext,
+    const WSK_CLIENT_CONNECTION_DISPATCH * Dispatch,
+    PEPROCESS OwningProcess,
+    PETHREAD OwningThread,
+    PSECURITY_DESCRIPTOR SecurityDescriptor,
+    PIRP Irp)
+{
+    UNIMPLEMENTED;
+    return STATUS_NOT_IMPLEMENTED;
+}
+
+static NTSTATUS WSKAPI
+WskControlClient(
+    _In_ PWSK_CLIENT Client,
+    _In_ ULONG ControlCode,
+    _In_ SIZE_T InputSize,
+    _In_reads_bytes_opt_(InputSize) PVOID InputBuffer,
+    _In_ SIZE_T OutputSize,
+    _Out_writes_bytes_opt_(OutputSize) PVOID OutputBuffer,
+    _Out_opt_ SIZE_T * OutputSizeReturned, _Inout_opt_ PIRP Irp)
+{
+    UNIMPLEMENTED;
+    return STATUS_NOT_IMPLEMENTED;
+}
+
+static struct _WSK_PROVIDER_DATAGRAM_DISPATCH UdpDispatch = {
+    .WskControlSocket = WskControlSocket,
+    .WskCloseSocket = WskCloseSocket,
+    .WskBind = WskBind,
+    .WskSendTo = WskSendTo,
+    .WskReceiveFrom = WskReceiveFrom,
+    .WskRelease = WskReleaseUdp,
+    .WskGetLocalAddress = WskGetLocalAddress,
+};
+
+/* Connection oriented routines (TCP/IP): */
+
+static NTSTATUS WSKAPI
+WskConnect(_In_ PWSK_SOCKET Socket, _In_ PSOCKADDR RemoteAddress, _Reserved_ ULONG Flags, _Inout_ PIRP Irp)
+{
+    PTDI_CONNECTION_INFORMATION TargetConnectionInfo, Ignored;
+    PIRP tdiIrp;
+    struct _WSK_SOCKET_INTERNAL *s = (struct _WSK_SOCKET_INTERNAL *)Socket;
+    NTSTATUS status;
+    struct NetioContext *nc;
+
+    IoSetNextIrpStackLocation(Irp);
+
+    status = STATUS_INVALID_PARAMETER;
+    if (s->LocalAddressHandle == NULL)
+    {
+        DbgPrint("LocalAddressHandle is not set, need to bind your socket before connecting\n");
+        goto err_out;
+    }
+
+    status = STATUS_INSUFFICIENT_RESOURCES;
+    nc = ExAllocatePoolWithTag(NonPagedPool, sizeof(*nc), 'NEIO');
+    if (nc == NULL)
+    {
+        goto err_out;
+    }
+    nc->socket = s;
+    nc->UserIrp = Irp;
+
+    tdiIrp = NULL;
+
+    status = TdiAssociateAddressFile(s->LocalAddressHandle, s->ConnectionFile);
+    if (!NT_SUCCESS(status))
+    {
+        goto err_out_free_nc;
+    }
+
+    TargetConnectionInfo = TdiConnectionInfoFromSocketAddress(RemoteAddress);
+    if (TargetConnectionInfo == NULL)
+    {
+        goto err_out_free_nc;
+    }
+
+    /* TODO: @thomasfabber: is this correct? */
+    Ignored = TdiConnectionInfoFromSocketAddress(RemoteAddress);
+    if (Ignored == NULL)
+    {
+        goto err_out_free_nc_and_tci;
+    }
+
+    IoMarkIrpPending(Irp);
+    SocketGet(s);
+
+    status = TdiConnect(&tdiIrp, s->ConnectionFile, TargetConnectionInfo, Ignored, NetioComplete, nc);
+
+    /* If allocating tdiIrp fails we get here.
+     * Call the IoCompletion of the application's Irp so this Irp
+     * gets freed.
+     */
+    if (!NT_SUCCESS(status))
+    {
+        SocketPut(s);
+        goto err_out_free_nc_and_tci;
+    }
+    return STATUS_PENDING;
+
+err_out_free_nc_and_tci:
+    ExFreePool(TargetConnectionInfo);
```

Comment:
Is it possible to use ExFreePoolWithTag with a suitable tag?

Reply:
Yes :)